tokio-util/src/time/delay_queue.rs

```diff
@@ -152,6 +160,256 @@ pub struct DelayQueue<T> {
+        for (old_key, remapped_key) in key_map.iter() {
+            inverse_key_map.insert(*remapped_key, *old_key);
+        }
+
+        let keys_given_out: HashSet<_> = keys_in_slab
+            .iter()
+            .map(|idx| {
+                let key = Key::new(*idx);
+
+                match inverse_key_map.get(&key.into()) {
+                    Some(remapped) => *remapped,
+                    None => key,
+                }
+            })
+            .collect();
+
+        let mut remapped_keys = HashSet::new();
+
+        slab.compact(|_, from, to| {
+            if let Some(old_key) = inverse_key_map.get(&KeyInternal::new(from)) {
+                // We previously re-mapped old_key -> KeyInternal(from) and now have to
+                // update the internal key since it has been re-mapped again to KeyInternal(to).
+                *key_map.get_mut(&old_key).unwrap() = KeyInternal::new(to);
+
+                remapped_keys.insert(*old_key);
+            } else {
+                let key = Key::new(from);
+                assert!(!key_map.contains_key(&key));
+
+                key_map.insert(Key::new(from), KeyInternal::new(to));
+                remapped_keys.insert(key);
+            }
+
+            true
+        });
+
+        // get all keys that we need to map to itself in `key_map`, these are
+        // all keys in `keys_given_out` that weren't re-mapped.
+        // See comment for `remove` method for why this identity mapping is necessary.
+        for key in keys_given_out.difference(&remapped_keys) {
+            assert!(!self.key_map.contains_key(key));
+            self.key_map.insert(*key, (*key).into());
+        }
+    }
+
+    // Re-maps a `Key` that was given out to the user to its corresponding internal key
+    fn remap_key(&self, key: &Key) -> KeyInternal {
+        let key_map = &self.key_map;
+        let remapped_key = match key_map.get(&*key) {
+            Some(k) => *k,
+            None => (*key).into(),
+        };
+        remapped_key
+    }
+
+    // We maintain a set of available keys for efficiency reasons, so as not to calculate
+    // the smallest available key each time `self.inner.insert` outputs a duplicate key.
+    // The creation of new keys is necessary if the `self.inner.insert` call
+    // in `self.insert` gives back a key that was previously given out.
+    // This scenario of a duplicate key can only happen after `compact` was called.
+    fn create_available_keys(&mut self) {
+        assert!(self.available_keys.is_empty());
+        self.available_keys.reserve(AVAILABLE_KEYS_SET_SIZE);
+
+        let mut i = 0;
+        let mut num_created_keys = 0;
+        while num_created_keys < AVAILABLE_KEYS_SET_SIZE {
+            if !self.key_map.contains_key(&Key::new(i)) {
+                self.available_keys.insert(KeyInternal::new(i));
+                num_created_keys += 1;
+            }
+            i += 1;
+        }
+    }
```

Comment:
@Darksonn I don't see a better solution to this. We can add  removed `Key`s back into `available_keys`, this slightly limits the number of `create_new_key` calls, but when we add many new keys after a `compact` call then I think we just have to live with a linear runtime every `AVAILABLE_KEYS_SET_SIZE` `insert` calls.

I'm not sure if/how bug prone simply incrementing an index to create new Keys would be in practice... I guess not really. But if the key index does overflow this way then the bug is really subtle. We can't really handle an overflow since we then have to check each new Key index again for duplication.

Do you have a solution for this?

Reply:
Having a counter for the next id seems fine to me if you do it like this:
```rs
while key_is_in_use(self.next_key) {
    self.next_key = self.next_key.wrapping_add(1);
}
return self.next_key;
```